### PR TITLE
mgr/dashboard: Config options integration (read-only) depends on #22422

### PR DESCRIPTION
**Pull request by s0nea**
This PR adds the config options stored by the MON database to the configuration documentation page. One can filter for these config options by setting the 'Source' filter to 'mon' on the configuration documentation page.

Related tests are also included.

The PR depends on https://github.com/ceph/ceph/pull/22422.

Signed-off-by: Tatjana Dehler <tdehler@suse.com>

### CODE REVIEW

File: qa/tasks/mgr/dashboard/test_config_options.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+from .helper import DashboardTestCase, authenticate
+
+
+class ConfigOptionsTest(DashboardTestCase):
+
+    @authenticate
+    def test_get_db_config_options(self):
+        data = self._get('/api/config_options')
+        self.assertStatus(200)
+        self.assertEqual(isinstance(data, list), True)
```
Comment (sebastian-philipp) [inline qa/tasks/mgr/dashboard/test_config_options.py]: You can also use 
```python
self.assertSchema(JList(sub_element=...))
```
here, if you want to.

File: qa/tasks/mgr/dashboard/test_config_options.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    @authenticate
+    def test_get_daemon_runtime_config(self):
+        def _get_func():
```
Comment (sebastian-philipp) [inline qa/tasks/mgr/dashboard/test_config_options.py]: What about calling this `get_mon_allow_pool_delete()`?

File: src/pybind/mgr/dashboard/HACKING.rst [hunk 3]
```diff
@@ -4,7 +4,7 @@ Dashboard Developer Documentation
 Frontend Development
 --------------------
 
-Before you can start the dashboard from within a development environment,  you
```
Comment (sebastian-philipp) [inline src/pybind/mgr/dashboard/HACKING.rst]: Is this related to this PR?

File: qa/tasks/mgr/dashboard/test_config_options.py (hunk 1 above)
Comment (votdev) [reply qa/tasks/mgr/dashboard/test_config_options.py]: What about ``self.assertIsInstance(data, list)`` ?

File: src/pybind/mgr/dashboard/HACKING.rst (hunk 3 above)
Comment (s0nea) [reply src/pybind/mgr/dashboard/HACKING.rst]: I only removed the unnecessary blank. Should I open a separate pull request for that?

Comment (s0nea) [reply src/pybind/mgr/dashboard/HACKING.rst]: I only removed the unnecessary blank. Should I open a separate pull request for that?

File: qa/tasks/mgr/dashboard/test_config_options.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        self._ceph_cmd(['config', 'set', 'mon', 'mon_allow_pool_delete', 'true'])
+        result = self._wait_for_expected_get_result(_get_mon_allow_pool_delete_config, 'true')
+        self.assertEqual(result, 'true')
+
+        self._ceph_cmd(['config', 'set', 'mon', 'mon_allow_pool_delete', 'false'])
+        result = self._wait_for_expected_get_result(_get_mon_allow_pool_delete_config, 'false')
+        self.assertEqual(result, 'false')
+
+    def _wait_for_expected_get_result(self, get_func, expected_result, max_attempts=3,
+                                      sleep_time=10):
```
Comment (sebastian-philipp) [inline qa/tasks/mgr/dashboard/test_config_options.py]: could you please only specify the timeout or the number of attepts (e.g. 30 seconds or attempts) and then try again every second?

File: qa/tasks/mgr/dashboard/test_config_options.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+from .helper import DashboardTestCase, authenticate
+
+
+class ConfigOptionsTest(DashboardTestCase):
+
+    @authenticate
+    def test_get_db_config_options(self):
+        data = self._get('/api/config_options')
+        self.assertStatus(200)
+        self.assertIsInstance(data, list)
```
Comment (sebastian-philipp) [inline qa/tasks/mgr/dashboard/test_config_options.py]: Could you also please verify the structure of the result?

**Comment by s0nea** [comment]
@ricardoasmarques oh, thanks. I will have a look at it :)

**Comment by s0nea** [comment]
> Could you run npx pretty-quick? This will format the .ts file.
@tspmelo Done. I'm not sure about line 73-80 in configuration.component.ts. It looks a bit strange now?

File: src/pybind/mgr/dashboard/controllers/cluster_configuration.py [hunk 6]
```diff
@@ -4,19 +4,42 @@
...
 from .. import mgr
+from ..services.ceph_service import CephService
 from . import ApiController, RESTController, AuthRequired
 
 
 @ApiController('/cluster_conf')
 @AuthRequired()
 class ClusterConfiguration(RESTController):
+
+    def _append_config_option_values(self, options):
```
Comment (p-na) [inline src/pybind/mgr/dashboard/controllers/cluster_configuration.py]: Using `enumerate` may simplify the function:

```python
    def _append_config_option_values(self, options):
        """ docs..."""
        config_dump = CephService.send_command('mon', 'config dump')
        for config_dump_entry in config_dump:
            for i, elem in enumerate(options):
                if config_dump_entry['name'] == elem['name']:
                    if 'value' not in elem:
                        options[i]['value'] = []
                        options[i]['source'] = 'mon'
                        
                    options[i]['value'].append({'section': config_dump_entry['section'],
                                                'value': config_dump_entry['value']})
        return options
```

File: qa/tasks/mgr/dashboard/test_cluster_configuration.py [hunk 7]
```diff
@@ -11,7 +13,8 @@ def test_list(self):
         self.assertIsInstance(data, list)
         self.assertGreater(len(data), 1000)
         for conf in data:
-            self._validate_single(conf)
+            contains_value = True if 'value' in conf else False
```
Comment (p-na) [inline qa/tasks/mgr/dashboard/test_cluster_configuration.py]: Could also be simplified to

```python
contains_value = 'value' in conf
```

**Comment by callithea** [comment]
Teuthology tests passed: http://pulpito.ceph.com/laura-2018-06-13_21:11:56-rados:mgr-wip-lpaduano-testing-distro-basic-smithi/

File: src/pybind/mgr/dashboard/HACKING.rst (hunk 3 above)
Comment (LenzGr) [reply src/pybind/mgr/dashboard/HACKING.rst]: I think it's fine to leave it in here.
